src/librustc/traits/select.rs

```diff
@@ -3293,6 +3293,17 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                     predicate: predicate.value
                 }))
         }).collect();
+        if predicates.len() > 1 {
```

Comment:
This is not only quadratic but also breaks the order assumption. We need to locate the root cause.

Reply:
I don't quite know what you're saying here. However, you could instead de-duplicate by accumulating into a hashset:

```rust
let mut hashset = HashSet::new();
predicates.retain(|&i| hashset.insert(i));
```

This is -- in theory -- O(1) and preserves ordering.

Reply:
Another option is:

```rust
predicates.sort_unstable(); 
predicates.dedup();
```

Reply:
No. First of all this doesn't preserve order, and second implement Ord on all transitively contained types is a frustrating thing.